Configuration manager non explorer version

h-kataria opened the pull request:
New screens under Configuration main tab with non-explorer version of
- Providers
- Profiles
- Configured Systems

These screens replace subtab `Management` under `Configuration` maintab.

@skateman it's a work in progress to address https://github.com/ManageIQ/manageiq-ui-classic/pull/6756#issuecomm. just creating a wip PR as i am working thru further cleanup and fixing things as i compare new screens with original explorer version to match all the current feature implementations. I will also add some screesnhots, once i have this bit more cleaned up.

Pending core https://github.com/ManageIQ/manageiq/pull/19949
Schema PR https://github.com/ManageIQ/manageiq-schema/pull/464

before changes screenshots
Sub menu under Configuration main menu
![before_cm_explorer](https://user-images.githubusercontent.com/3450808/77481815-da4dba00-6dfa-11ea-85ab-76d6b41e2648.png)

Configuration Manager node selection
![cm_node_summary](https://user-images.githubusercontent.com/3450808/77481824-dde14100-6dfa-11ea-8675-b87c5cab01ce.png)

Configuration Profile node selection
![cp_node_summary](https://user-images.githubusercontent.com/3450808/77481827-e174c800-6dfa-11ea-8c1e-edf91a63a5df.png)

Configured System node selection
![cs_node_summary](https://user-images.githubusercontent.com/3450808/77481831-e33e8b80-6dfa-11ea-92d6-6e91fa3b5cb9.png)

after changes screenshots
New sub menus under Configuration main menu
![configuration_menu](https://user-images.githubusercontent.com/3450808/77480926-049e7800-6df9-11ea-981e-17d74dca2086.png)

Blank screen with Add Provider button where there are no Configuration Manager Providers yet in db
![after](https://user-images.githubusercontent.com/3450808/79883260-b1045780-83c1-11ea-9248-1123f8020fa1.png)

List of all Configuration Managers with advanced search/filters
![cm_list](https://user-images.githubusercontent.com/3450808/77542082-05c4b900-6e7c-11ea-800c-e03fcbdb8738.png)

Configuration Manager Summary
![cm_summary](https://user-images.githubusercontent.com/3450808/77545283-97cec080-6e80-11ea-8e5e-3ec7703ffcb7.png)

List of All Configuration Profiles with advanced search/filters
![Screenshot from 2020-04-14 17-45-06](https://user-images.githubusercontent.com/3450808/79277413-c1737a00-7e77-11ea-81cf-1ba2f8367c1a.png)

Configuration Profile Summary
![cp_summary](https://user-images.githubusercontent.com/3450808/77545275-943b3980-6e80-11ea-80bf-206f487b40c8.png)

List of All Configured Systems
![cs_list_view](https://user-images.githubusercontent.com/3450808/77481113-5f37d400-6df9-11ea-8d5e-3d4fcf875ba0.png)

Configured System Summary
![cs_summary](https://user-images.githubusercontent.com/3450808/77545261-90a7b280-6e80-11ea-8c05-0cccc9c741b9.png)

skateman commented:
Looks great on the screenshots, I think we should support filters on the two empty sidebars as well. I'm not sure how complex it is to do so, so it might be better to do it in a separate PR. Anyway, great job :+1:

h-kataria commented:
> Looks great on the screenshots, I think we should support filters on the two empty sidebars as well. I'm not sure how complex it is to do so, so it might be better to do it in a separate PR. Anyway, great job

That's an easy one to do, i will add that in here

skateman commented on app/helpers/application_helper.rb:
```diff
@@ -648,14 +648,6 @@ def taskbar_in_header?
-    ]
-    nameonly_search_layouts.include?(@layout)
-  end
-
```
:heart_eyes:

skateman commented on app/controllers/application_controller/ci_processing.rb:
```diff
@@ -143,7 +143,7 @@ def process_elements(elements, klass, task, display_name = nil, order_field = ni
 
   def explorer_controller?
-    %w[vm_cloud vm_infra vm_or_template infra_networking].include?(controller_name)
+    %w[vm_cloud vm_infra vm_or_template infra_networking automation_manager].include?(controller_name)
```
shouldn't be `configuration_manager`?

h-kataria replied:
no, `configuration_manger` is no longer an explorer controller but shares some code with `automation_manager` controller and that is explorer style screen, so i need to add this here so i can redirect appropriately form shared methods.

skateman replied:
OH, right

skateman commented:
Tried, seems working, @mzazrivec wanna test too?

himdel commented:
Oh, I just merged https://github.com/ManageIQ/manageiq-ui-classic/pull/6815, hence the conflict...

@h-kataria The non-explorer version will need a toolbar change...

```diff
- :tree_select    => 'root',
+ :redirect_url   => '/the_controller/show_list'
```

h-kataria commented:
@mzazrivec can you please test this PR in UI.

chessbyte commented:
@mzazrivec any luck verifying this?

himdel commented on app/controllers/configuration_profile_controller.rb:
```diff
@@ -0,0 +1,51 @@
+  after_action :set_session_data
+
+  def self.table_name
+    @table_name ||= "configuration_profile"
```
We should be able to replace these with

```
def self.model
  ConfigurationProfile
end
```

(`table_name` defaults to `model.name.underscore`)

(technically, even the model method is not needed as the default implementation will guess from the controller name, but... :))

himdel commented on app/controllers/configuration_profile_controller.rb:
```diff
@@ -0,0 +1,51 @@
+class ConfigurationProfileController < ApplicationController
+  include Mixins::GenericListMixin
```
Missing the standard..

```
  before_action :check_privileges
  before_action :get_session_data
  after_action :cleanup_action
  after_action :set_session_data
```

himdel replied:
Oh never mind, just different ordering.

himdel commented on app/controllers/configuration_profile_controller.rb:
```diff
@@ -0,0 +1,51 @@
+
+  def button
+    @edit = session[:edit] # Restore @edit for adv search box
+    params[:display] = @display if %w[configured_systems].include?(@display) # Were we displaying nested list
```
sounds like this should be `if display_methods.include?(@display)`

himdel commented on app/controllers/ems_configuration_controller.rb:
```diff
@@ -0,0 +1,103 @@
+
+  def button
+    @edit = session[:edit] # Restore @edit for adv search box
+    params[:display] = @display if %w[configured_systems].include?(@display) # Were we displaying nested list
```
same here (`if display_methods.include?(@display)`), but what happens for nested lists of configuration_profiles?

himdel commented on app/controllers/configured_system_controller.rb:
```diff
@@ -0,0 +1,46 @@
+    tag(model) if params[:pressed] == "configured_system_tag"
+    provision if params[:pressed] == "configured_system_provision"
+
+    return if ["#{params[:controller]}_tag"].include?(params[:pressed]) && @flash_array.nil? # Tag screen showing
```
this line doesn't do anything, early return on the last line of a function

himdel commented on app/controllers/ems_configuration_controller.rb:
```diff
@@ -0,0 +1,103 @@
+      delete
+    end
+
+    return if ["#{params[:controller]}_tag"].include?(params[:pressed]) && @flash_array.nil? # Tag screen showing
```
should we move this line right after the `tag(mode)` line? as both happen under the same button..

h-kataria commented:
@himdel made suggested changes and fixed couple of things found during testing.
- added missing route while trying to Apply tags on nested list of Configured system thru Configuration Profiles relationship
- do not show checkboxes on list of Configuration Profiles as that doesnt have any toolbar buttons

himdel commented:
Thanks, I don't see any issues with the code,
but I still want to go through the toolbars to make sure the explorer/nonexplorer differences didn't break any forms/buttons there :)

himdel commented:
BTW why is travis still red?

Should I wait with final reviews until the core PR is merged? Or does this have to go in at the same time?

Looks like a *lot* of RBAC renames to do at the same time as another UI rewrite.

h-kataria commented:
> BTW why is travis still red?
> 
> Should I wait with final reviews until the core PR is merged? Or does this have to go in at the same time?
> 
> Looks like a _lot_ of RBAC renames to do at the same time as another UI rewrite.
> EDIT: but I see the table in [ManageIQ/manageiq#19949 (comment)](https://github.com/ManageIQ/manageiq/pull/19949#issuecomment-613474884), makes sense

Fixing merge conflict now, and this should be merged along with Core/Schema PR.

h-kataria commented:
@himdel travis failure should be green one https://github.com/ManageIQ/manageiq/pull/19949 is merged

himdel commented:
@h-kataria I think you missed one toolbar...

```
app/helpers/application_helper/toolbar/configuration_scripts/policy_mixin.rb
3:    included_class.button_group('provider_foreman_policy',
```

himdel commented:
In Configuration > Configured Systems,

I see a lot of Configured System (Ansible Tower), despite having no providers in Configuration.
Sounds like we should be filtering out ansible tower systems?

Unless the plan is to merge Ansible into Configuration.

himdel commented:
Configuration > Configured Systems,
Lifecycle > Provision..

always returns a 204 response for me (could be related to the wrong configured system type)

himdel commented:
(Can't test the rest without an accessible foreman instance, waiting for that.)

himdel commented:
Adding a foreman provider:

Started `POST "/ems_configuration/edit/new?button=save"` for ::1 at 2020-04-16 13:16:36 +0000
Processing by `EmsConfigurationController#edit` as JS
Parameters: `{"name"=>"test", "zone"=>"default", "url"=> "...", "default_userid"=>"...", "default_password"=>"[FILTERED]", "default_auth_status"=>"on", "button"=>"save", "id"=>"new"}`

Error caught: [NameError] undefined local variable or method `construct_edit_for_audit' for #<EmsConfigurationController:0x00005647cb6ea118>
/home/himdel/manageiq-ui-classic/app/controllers/mixins/manager_controller_mixin.rb:38:in `save_provider'
/home/himdel/manageiq-ui-classic/app/controllers/mixins/manager_controller_mixin.rb:112:in `edit'

h-kataria commented:
> In Configuration > Configured Systems,
> 
> I see a lot of Configured System (Ansible Tower), despite having no providers in Configuration.
> Sounds like we should be filtering out ansible tower systems?
> 
> Unless the plan is to merge Ansible into Configuration.
> In which case the textual summary is wrong because it now always creates a parent link to `ems_configuration`, so the link goes to a "Can't access selected records" screen.

@fryguy @bdunne what query should i use to filter out `Ansible Tower` related records from the new screens

himdel commented:
Deleting a provider from the detail screen seems to try to do treeSelect, but that's a bug from https://github.com/ManageIQ/manageiq-ui-classic/pull/6815, the toolbar definition is right, fixing in https://github.com/ManageIQ/manageiq-ui-classic/pull/6984

h-kataria commented:
> Configuration > Configured Systems,
> Lifecycle > Provision..
> 
> always returns a 204 (No Content) response for me (could be related to the wrong configured system type)

it works for me when i select Foreman configured system but breaks for Ansible Tower configured system. Let's see if we decide to filter those records out from the list.

Addressed other two comments, about typo in toolbar and error during save of Provider

h-kataria commented:
> Configuration > Configured Systems,
> Lifecycle > Provision..
> 
> always returns a 204 (No Content) response for me (could be related to the wrong configured system type)

@himdel fixed this issue.
![Screenshot from 2020-04-18 21-41-42](https://user-images.githubusercontent.com/3450808/79677324-05d78080-81be-11ea-806a-e41bbd2b7e0b.png)

h-kataria commented:
> In Configuration > Configured Systems,
> 
> I see a lot of Configured System (Ansible Tower), despite having no providers in Configuration.
> Sounds like we should be filtering out ansible tower systems?
> 
> Unless the plan is to merge Ansible into Configuration.
> In which case the textual summary is wrong because it now always creates a parent link to `ems_configuration`, so the link goes to a "Can't access selected records" screen.

@himdel @fryguy @bdunne Added a named_scope to exclude Ansible Tower configured systems from the list, please review

before
![before_named_scope](https://user-images.githubusercontent.com/3450808/79677475-ee00fc00-81bf-11ea-8a37-3f43c2192e7f.png)

after
![after_named_scope](https://user-images.githubusercontent.com/3450808/79677481-f3f6dd00-81bf-11ea-8624-f9542177c2de.png)

himdel commented:
Adding a provider works without errors :+1:,
I can only see the right kind of configured systems,
no 204, a flash message.

Looks like everything works now :)

himdel commented:
One thing I notice (model side issue):

* add a foreman provider,
* remove it
* an Endpoint still remains
* => try to add the same provider again
* fail because Endpoint.url must be unique, and there is already such an endpoint

Cc @bdunne maybe? :)

skateman commented:
@himdel just to be clear, the issue happens only with this PR or also on master?

bdunne commented:
> One thing I notice (model side issue I think):
> 
>     * add a foreman provider,
> 
>     * remove it
> 
>     * an Endpoint still remains
> 
>     * => try to add the same provider again
> 
>     * fail because Endpoint.url must be unique, and there is already such an endpoint
> 
>     * (so I can add the provider max 4 times by using hostname or ip, with or without final slash)
> 
> 
> Cc @bdunne maybe? :)

[The relations are :dependent => :destroy](https://github.com/ManageIQ/manageiq-providers-foreman/blob/233f6075e7714a9d3162c21dd1356b61ebd6bbe7/app/models/manageiq/providers/foreman/provider.rb#L13) so I would expect everything to be cleaned up if you're calling `#destroy` on the provider.

It appears to work for me.  `ManageIQ::Providers::Foreman::Provider.create!` creates 1 Provider, 2 ExtManagementSystems and 1 Endpoint.  Calling destroy on that Provider removes all of those records.

himdel commented:
@bdunne You may be missing the point. We never remove the Provider. We remove the Manager. And that destroys nothing.

Fryguy commented:
cc @agrare ^?   I know we do something with other providers that have a literal Provider.

bdunne commented:
Oh, well then that explains why there is stuff left in the DB.  Can it be changed to remove the Provider instead?

Fryguy commented on app/controllers/ems_configuration_controller.rb:
```diff
@@ -0,0 +1,106 @@
+  end
+
+  def provider_class
+    ManageIQ::Providers::Foreman::Provider
```
Why is Foreman still here?

h-kataria replied:
@Fryguy This is still here because currently UI does not have a  type drop down in Add a new Provider form, i was hoping once we have new Data driven Provider form in place we will have a type drop down for user to select what type of Provider they want to add with their respective fields. Do you think i should i add a Type drop down on the Add form, if yes, what should be the options, then we can change this code to use value from the type field.

Fryguy replied:
How do we do it for other providers?

Fryguy replied:
Or is this related to the problem with the Provider model vs ExtManagementSystem model?

h-kataria replied:
other provider have a type field on the form, based upon the type selected a Provider object is created. I am working on adding a `Type` field to the Add Configuration Manager form with similar logic that we have for other providers. Should get that in shortly.

h-kataria replied:
@Fryguy I am running into an issue with type form field, I have code implemented the same way as it is in other providers.

I am using `EmsConfiguration.supported_types_and_descriptions_hash` to build Type drop down on Configuration Manager Add form, that returns me `cam_configuration` and `foreman_configuration`
but trying to save the new record it throws an error `Error caught: [ActiveModel::MissingAttributeError] can't write unknown attribute "url"`. 
When Add button is pressed in UI, I am creating a new Provider object `EmsConfiguration.model_from_emstype("foreman_configuration").new` that returns me an object of type `ManageIQ::Providers::Foreman::ConfigurationManager ` which does not have url attribute, before my changes we had this part hardcoded to create an object of type `ManageIQ::Providers::Foreman::Provider` 
Please advice.

h-kataria replied:
I was able to resolve the issue i mentioned above by adding `supported_types_and_descriptions_hash` method to Provider class but i am not sure if that's the correct fix
```
  def self.supported_types_and_descriptions_hash
    supported_subclasses.each_with_object({}) do |klass, hash|
      if Vmdb::PermissionStores.instance.supported_ems_type?(klass.ems_type)
        hash[klass.ems_type] = klass.description
      end
    end
  end
```

Fryguy replied:
@agrare ^?

What do you think?  (we need some sort of Providers permissions I think, or perhaps a way to get the permitted managers, and then from that get their Providers?)

agrare replied:
Particularly for something like Foreman where neither Manager is the "primary" then permissions on the Provider level are the only way I think.

h-kataria replied:
@agrare @Fryguy can you guys help with this change, I am not sure I understand where the permissions need to be set or what other change needs to be made in models.

himdel replied:
Is there a point to all this?

Consider that for now Foreman::Provider is precisely the class this controller needs.

None of this code should exists in the first place, aren't we just wasting time that could be spent on finishing the proper way of doing that via the API?

h-kataria replied:
@Fryguy I agree with @himdel that's the the reason i did not change Add new provider form in this PR. Even if we decide to make changes to existing form , i would prefer a separate follow up PR, this PR is getting way too big.

agrare replied:
@h-kataria happy to help add permissions and supported_for_create to the provider, would rather that than hard coding Foreman

Fryguy replied:
If these are meant to be temporary, then I'm fine with that, especially as the goal is to move towards pluggable DDF for this feature.

> aren't we just wasting time that could be spent on finishing the proper way of doing that via the API?

Isn't that what we are doing with this discussion? It's clear there's a general design
issue with respect to provider/manager split, so discussing it is something we need to do
anyway for the API. It helps to have that discussion in context with the code that is
actually consuming it, hence me bringing in @agrare, as he has some opinions on how best
to deal with the provider/manager split.
Also this thread started as me noticing Foreman in the code when the purpose of the PR is
to remove Foreman from the code, so it stood out. It's not clear from the code that the
purpose for the method is to drive the drop-down for adding a new provider.

h-kataria replied:
line of code in question does not drive the drop-down as currently there is no Type drop-down in the UI. Currently users can only add a Foreman Provider and this line is providing class for the new record, when the Add button is UI is pressed.

Fryguy commented on app/controllers/mixins/manager_controller_mixin.rb:
```diff
@@ -94,126 +84,25 @@ def authentication_validate
+        render_form
+      else
+        @in_a_form = true
+        @redirect_controller = "ems_configuration"
```
This seems weird in a general ManagerControllerMixin

Fryguy commented on app/controllers/mixins/manager_controller_mixin.rb:
```diff
@@ -222,20 +111,26 @@ def edit
+          render_form
+        else
+          @in_a_form = true
+          @redirect_controller = "ems_configuration"
```
Ditto

Fryguy commented on app/controllers/restful_redirect_controller.rb:
```diff
@@ -6,9 +6,7 @@ def index
-        if %w[ManageIQ::Providers::ConfigurationManager].include?(record.type) || record.type.starts_with?('ManageIQ::Providers::Foreman')
-          redirect_to(:controller => 'provider_foreman', :action => 'show', :id => params[:id])
-        elsif %w[ManageIQ::Providers::AnsibleTower::AutomationManager].include?(record.type)
+        if %w[ManageIQ::Providers::AnsibleTower::AutomationManager].include?(record.type)
```
Can we prefer `if record.kind_of?(ManageIQ::Providers::ExternalAutomationManager)`, so we can remove the plugin reference entirely?

Fryguy replied:
I see technically this was original code (didn't realize initially), so this can be a followup

Fryguy commented on app/controllers/restful_redirect_controller.rb:
```diff
@@ -6,9 +6,7 @@ def index
-        elsif %w[ManageIQ::Providers::AnsibleTower::AutomationManager].include?(record.type)
+        if %w[ManageIQ::Providers::AnsibleTower::AutomationManager].include?(record.type)
           redirect_to(:controller => 'automation_manager', :action => 'show', :id => params[:id])
         elsif %w[ManageIQ::Providers::EmbeddedAnsible::AutomationManager].include?(record.type)
```
Same here - `if record.kind_of?(ManageIQ::Providers::EmbeddedAutomationManager)`

Fryguy replied:
I see technically this was original code (didn't realize initially), so this can be a followup

Fryguy commented on app/helpers/application_helper/toolbar/ems_configurations_center.rb:
```diff
@@ -1,43 +1,40 @@
+  button_group('ems_configuration_vmdb', [
     select(
-      :provider_vmdb_choice,
+      :ems_configuration__vmdb_choice,
```
2 underscores?

Fryguy commented on spec/controllers/application_controller/advanced_search_spec.rb:
```diff
@@ -1,18 +1,20 @@
-    stub_user(:features => :all)
-    controller.instance_variable_set(:@sb, {})
-  end
+describe EmsConfigurationController, "::AdvancedSearch" do
```
Not for this PR, but it's weird to me to have a `EmsConfigurationController, "::AdvancedSearch"` in a file named `application_controller/advanced_search_spec.rb ` as the latter is generic and the former is specific.

Fryguy commented on spec/controllers/restful_redirect_controller_spec.rb:
```diff
@@ -44,12 +44,6 @@
       expect(response).to redirect_to(:controller => "vm_cloud", :action => "show", :id => vm_cloud.id)
     end
 
-    it 'redirects to provider_foreman controller' do
```
did we lose this test for for configuration managers? I don't see elsewhere in this code, but maybe I just missed it.

h-kataria replied:
good catch @Fryguy that test was removed because i removed https://github.com/ManageIQ/manageiq-ui-classic/pull/6782/files#diff-54afc0af5eff6d605e58aea868f5db92L9-L10 from controller but it appears we do no need that code, so i will adjust and put the test and controller code back in

Fryguy commented on spec/presenters/tree_node/ext_management_system_spec.rb:
```diff
@@ -28,7 +28,6 @@
     :ems_vmware_cloud                 => { :tip_prefix => 'Ems Cloud' },
     # Other remaining providers
     :automation_manager_ansible_tower => { :key_prefix => 'at-' },
-    :configuration_manager_foreman    => { :key_prefix => 'fr-' },
```
should the configuration_manager entry about 12 lines down be uncommented and/or deleted?

Fryguy replied:
@h-kataria Is this a valid concern? ☝️

h-kataria replied:
@Fryguy now that the tree is removed from the left side in Configuration Managers area, we do not need this entry.

Fryguy commented on app/controllers/mixins/automation_manager_controller_mixin.rb:
```diff
@@ -0,0 +1,363 @@
+    end
+
+    def breadcrumb_name(_model)
+      "#{ui_lookup(:ui_title => 'foreman')} #{_('Provider')}"
```
Is ui_title correct here (i.e. is it still 'foreman')?

simaishi commented:
Jansa backport details:

```
$ git log -1
commit 978f54bfd2a26a3802d9181de86a8841c3cb26ad
Author: Jason Frey <fryguy9@gmail.com>
Date:   Thu May 7 22:06:12 2020 -0400

    Merge pull request #6782 from h-kataria/configuration_manager_non_explorer_version

    Configuration manager non explorer version

    (cherry picked from commit fdc0b7a2e3794dc15e077ff71e9841486f4c8524)
```
